File: exemplos/aula_250513/autenticacao/Autenticacao.java
package exemplos.aula_250513.autenticacao;

import java.util.List;

interface Autenticavel {
    public boolean autenticar(String senha);
}

interface CRAA{
    public double calcularCRAA();
}

abstract class Usuario {
    protected String nome;
}

class Professor extends Usuario implements Autenticavel {
    private String senha;

    public Professor(String nome, String senha) {
        this.nome = nome;
        this.senha = senha;
    }

    @Override
    public boolean autenticar(String senha) {
        return this.senha.equals(senha);
    }
}

class Aluno extends Usuario implements CRAA, Autenticavel {
    public String matricula;
    private String senha;

    public Aluno(String nome, String matricula) {
        this.nome = nome;
        this.matricula = matricula;
    }

    @Override
    public double calcularCRAA() {
        // Implementação do cálculo do CRAA
        return 0.0; // Exemplo de retorno
    }

    @Override
    public boolean autenticar(String senha) {
        return this.senha.equals(senha);
    }
}

class App {
    public static void main(String[] args) {
        Professor helder = new Professor("Hélder", "senha123");
        Aluno caua = new Aluno("Cauã", "123456");

        
        if (helder.autenticar("senha123")) {
            System.out.println("Professor autenticado com sucesso!");
        } else {
            System.out.println("Falha na autenticação do professor.");
        }

        System.out.println("Aluno + " + caua.nome + " matricula: " + caua.matricula); 

        CRAA jessica = new Aluno("Jéssica", "2023002");

        List<Autenticavel> lista = List.of(helder, caua);
    }
}

File: exemplos/aula_250513/autenticacao/autenticacao.py
from abc import ABC, abstractmethod

class IAutenticacao(ABC):
    @abstractmethod
    def autenticar(self, senha):
        pass
    
class ICRAA(ABC):
    @abstractmethod
    def calcularCRAA(self):
        pass
    
    
class Usuario():
    def __init__(self, nome):
        self.nome = nome
        
class Professor(Usuario, IAutenticacao):
    def __init__(self, nome, senha):
        super().__init__(nome)
        self.senha = senha
        
    def autenticar(self, senha):
        return self.senha == senha
    
class Aluno(Usuario, ICRAA, IAutenticacao):
    def __init__(self, nome, senha, notas):
        super().__init__(nome)
        self.senha = senha
        self.notas = notas
        
    def calcularCRAA(self):
        return sum(self.notas) / len(self.notas)
    
    def autenticar(self, senha):
        return self.senha == senha
    
# Exemplo de uso

professor = Professor("Carlos", "senha123")
aluno = Aluno("Ana", "senha456", [8.0, 9.0, 7.5])   
print(professor.autenticar("senha123"))  # True
print(aluno.autenticar("senha456"))  # True
print(aluno.calcularCRAA())  # 8.166666666666666 